assets/src/admin/paired-browsing/app.js

```diff
@@ -0,0 +1,215 @@
+				setTimeout( reject, 5000 );
+			} ),
+			new Promise( ( resolve, reject ) => {
+				this.ampIframe.addEventListener( 'load', resolve );
+				setTimeout( reject, 5000 );
+			} ),
+		];
+	}
+
+	/**
+	 * Validates whether or not the window document is AMP compatible.
+	 *
+	 * @param {Document} doc Window document.
+	 * @return {boolean} True if AMP compatible, false if not.
+	 */
+	documentIsAmp( doc ) {
+		return doc.documentElement.hasAttribute( 'amp' ) || doc.documentElement.hasAttribute( '⚡️' );
+	}
+
+	/**
+	 * Toggles the 'disconnected' class on the iframe, for whether or not the client in the iframe
+	 * is detected.
+	 */
+	checkConnectedIframes() {
+		this.ampIframe.classList.toggle(
+			'disconnected',
+			! ( this.nonAmpIframe.contentWindow && this.nonAmpIframe.contentWindow.ampPairedBrowsingClient ),
+		);
+
+		this.nonAmpIframe.classList.toggle(
+			'disconnected',
+			! ( this.ampIframe.contentWindow && this.ampIframe.contentWindow.ampPairedBrowsingClient ),
+		);
+	}
+
+	/**
+	 * Removes AMP related query variables from the supplied URL.
+	 *
+	 * @param {string} url URL string.
+	 * @return {string} Modified URL without any AMP related query variables.
+	 */
+	removeAmpQueryVars( url ) {
+		return removeQueryArgs( url, ampSlug, 'amp_validation_errors' );
+	}
+
+	/**
+	 * Adds the AMP query variable to the supplied URL.
+	 *
+	 * @param {string} url URL string.
+	 * @return {string} Modified URL with the AMP query variable.
+	 */
+	addAmpQueryVar( url ) {
+		return addQueryArgs(
+			url,
+			{
+				[ ampSlug ]: '',
+			},
+		);
+	}
+
+	/**
+	 * Adds the AMP paired browsing query variable to the supplied URL.
+	 *
+	 * @param {string} url URL string.
+	 * @return {string} Modified URL with the AMP paired browsing query variable.
+	 */
+	addPairedBrowsingQueryVar( url ) {
+		return addQueryArgs(
+			url,
+			{
+				[ ampPairedBrowsingQueryVar ]: '1',
+			},
+		);
+	}
+
+	/**
+	 * Removes the URL hash from the supplied URL.
+	 *
+	 * @param {string} url URL string.
+	 * @return {string} Modified URL without the hash.
+	 */
+	removeUrlHash( url ) {
+		const parsedUrl = new URL( url );
+		parsedUrl.hash = '';
+		return parsedUrl.href;
+	}
+
+	/**
+	 * Checks if a URL has the 'amp_validation_errors' query variable.
+	 *
+	 * @param {string} url URL string.
+	 * @return {boolean} True if such query var exists, false if not.
+	 */
+	urlHasValidationErrorQueryVar( url ) {
+		return hasQueryArg( url, 'amp_validation_errors' );
+	}
+
+	/**
+	 * Registers the provided client window with its parent, so that it can be managed by it.
+	 *
+	 * @param {Window} win Document window.
+	 */
+	registerClientWindow( win ) {
+		let oppositeWindow;
+
+		if ( win === this.ampIframe.contentWindow ) {
+			if ( ! this.documentIsAmp( win.document ) ) {
+				if ( this.urlHasValidationErrorQueryVar( win.location.href ) ) {
+					// eslint-disable-next-line no-alert
+					alert( 'The AMP version of this page could not be rendered due to validation errors.' );
```

Comment:
As noted in comment, let's find an alternative to using `alert()`. 

Aside: this needs to be translatable (in `__()`).